rust/worker/src/segment/record_segment.rs

```diff
@@ -268,13 +334,172 @@ impl RecordSegmentWriter {
+    #[error("Error setting to blockfile")]
+    BlockfileSetError,
+    #[error("Error deleting from blockfile")]
+    BlockfileDeleteError,
+    #[error("Embedding not set in the user write")]
+    EmbeddingNotSet,
+}
+
+impl ChromaError for ApplyMaterializedLogError {
+    fn code(&self) -> crate::errors::ErrorCodes {
+        match self {
+            ApplyMaterializedLogError::BlockfileSetError => ErrorCodes::Internal,
+            ApplyMaterializedLogError::BlockfileDeleteError => ErrorCodes::Internal,
+            ApplyMaterializedLogError::EmbeddingNotSet => ErrorCodes::InvalidArgument,
+        }
     }
+}
 
-    fn apply_log_chunk(&self, records: Chunk<LogRecord>) {
-        todo!()
+impl<'a> SegmentWriter<'a> for RecordSegmentWriter {
+    async fn apply_materialized_log_chunk(
+        &self,
+        records: Chunk<MaterializedLogRecord<'a>>,
+    ) -> Result<(), ApplyMaterializedLogError> {
+        for (log_record, _) in records.iter() {
+            match log_record.final_operation {
+                Operation::Add => {
+                    // Set all four.
+                    // Set user id to offset id.
+                    match self
+                        .user_id_to_id
+                        .as_ref()
+                        .unwrap()
+                        .set::<&str, u32>("", log_record.user_id.unwrap(), log_record.offset_id)
+                        .await
+                    {
+                        Ok(()) => (),
+                        Err(_) => {
+                            return Err(ApplyMaterializedLogError::BlockfileSetError);
+                        }
+                    };
+                    // Set offset id to user id.
+                    match self
+                        .id_to_user_id
+                        .as_ref()
+                        .unwrap()
+                        .set::<u32, &str>("", log_record.offset_id, log_record.user_id.unwrap())
+                        .await
+                    {
+                        Ok(()) => (),
+                        Err(_) => {
+                            return Err(ApplyMaterializedLogError::BlockfileSetError);
+                        }
+                    };
+                    // Set data record.
+                    match self
+                        .construct_and_set_data_record(
+                            log_record,
+                            log_record.user_id.unwrap(),
+                            log_record.offset_id,
+                        )
+                        .await
+                    {
+                        Ok(()) => (),
+                        Err(e) => {
+                            return Err(e);
+                        }
+                    }
+                    // Set max offset id.
+                    match self
+                        .max_offset_id
+                        .as_ref()
+                        .unwrap()
+                        .set("", MAX_OFFSET_ID, log_record.offset_id)
```

Comment:
is the log guaranteed to be in max offset id order?

Reply:
Summarizing our discussion here for posterity: This turned out to be a bug since the materializer is not guaranteed to return logs in the same order as writes. Changed the behavior to return sorted by offset logs. We also spent ~5 mins thinking about how we can improve this abstraction since readers don't need the offset id so it feels like there's some leaky abstraction here.